libs/GasBurnerControl/GasBurnerControl.h

```diff
@@ -0,0 +1,69 @@
+#ifndef GasBurnerControl_h
+#define GasBurnerControl_h
+
+
+#include "Arduino.h"
+
+
+/* OPTIONS */
+#define GBC_N_DEJAM_ATTEMPTS 3      // number of unsuccessful dejam attempts before aborting permanently (1-255)
+#define GBC_N_IGNITION_ATTEMPTS 3   // number of unsuccessful ignition attempts before aborting permanently (1-255)
+
+#define GBC_HIGH 0                  // logical high (0: GND, 1: VCC)
+#define GBC_LOW 1                   // logical low (0: GND, 1: VCC)
+
+#define GBC_SERIAL_DEBUG            // if uncommented every update will also send a message over serial
+
+
+/* TIMER SETTINGS */
+#define GBC_START_DELAY 2           // s, initial delay after powering the GBC
+#define GBC_DEJAM_DELAY 65          // s, wait time until dejamming is possible
+#define GBC_DEJAM_DELAY2 10         // s, wait time between additional dejam attempts
+#define GBC_IGNITION_DELAY 22       // s, time after which ignition should be complete
+#define GBC_DEJAM_DURATION 1000     // ms, duration of button press during dejamming
+#define GBC_POST_DEJAM_DELAY 1000   // ms, delay after dejam button release
+
+
+/* STATE REPRESENTATIONS (must be unique) */
+#define GBC_IDLE 0
+#define GBC_STARTING 1
+#define GBC_IGNITION 2
+#define GBC_RUNNING 3
+#define GBC_DEJAM 4
+#define GBC_ERROR 5
```

Comment:
I'd rather define an `enum class`, like

```cpp
enum class BurnerControlState {
    idle,
    starting,
    ignition,
    running,
    dejam
    error
};
```
Also, it might make sense to just use a `gbc` namespace instead of prefixing everything.

Reply:
For now I  have changed the states to `enum class` in 576044a.

At the moment I use:
```cpp
enum class GasBurnerControlState {
    idle = 0,
    starting = 1,
    ignition = 2,
    running = 3,
    dejam = 4,
    error = 5
};
```

In order to get rid of the GBC_SERIAL_DEBUG macro, I would like to expose more details in the state, ie:
* "substates" in dejam (preDejamDelay // pressButton // buttonIsPressend // releaseButton // postDejamDelay)
* "substates" for different types of errors

I was thinking about something like:
```cpp
enum class GasBurnerControlState {
    idle = 0,
    starting = 1,
    ignition = 2,
    running = 3,
    dejam_start = 40,
    dejam_preDejamDelay = 41,
    [etc.]
    error_ignition = 51,
    error_at_start = 52,
    error_wiring = 53
};
```

In the case of the `switch` this is fairly easy to group the substates
```cpp
switch (m_state) {
    case(dejam_start):
    case(dejam_preDejamDelay):
        //do stuff here
        break;
}
```

But what if I want to capture an error in the main loop and do not care which error it is?
```cpp
if (getState() in <ANY OF THE ERROR STATES>)
```
**How would I do that?**

Certainly this is not elegant:
```cpp
if(getState() >= 50)
```


> Also, it might make sense to just use a `gbc` namespace instead of prefixing everything. Then it could be even shorter, i.e. `gbc::State`.

I will have a look at that later.

Reply:
> Certainly this is not elegant

Hmm, maybe it really makes sense to have the error state here and the actual error condition stored in a separate enum?

Reply:
I gave it some thought: Knowing which sort of error occurred may be more relevant for the user. This is because "GFA never reached ignition state but exceeded dejam attempts" indicates that the power supply polarity is wrong, while "GFA tried ignition but never transitioned to running" implies that gas supply is off or flame watching sensor is poorly positioned. Hence, these cases require different manual actions.

At the same time I'd like `getState()` or equivalent include all the information relevant for the Arduino UI or the PI UI. Otherwise we had to expose the substates individually and recombine them in main app for transferring them to the Pi.

How about something like:
```cpp
enum class GasBurnerControlState {
    idle = 0,
    starting = 10,
    ignition = 20,
    running = 30,
    any_dejam = 4,   // never actually used in the state machine but for external use
    dejam_start = 40,
    dejam_preDejamDelay = 41,
    [etc.]
    any_error = 5,     // never actually used in the state machine but for external use
    error_ignition = 51,
    error_at_start = 52,
    error_wiring = 53
};
```

Then we could do `if(getState() / 10 == GasBurnerControlState::any_error)` Not pretty but it does the job. (I did not find a neat way to extract the most significant digit of an integer with arbitrary length. Saw stuff using `log(10`)

# Bonus

If we want the state to be fully complete, we would have to consider the dejam and ignition attempt counters as well. This is a bonus feature. In this case, a format easily pulled by the Pi with one command could be something like `getFullState()` returning:
```
2021
^^^^
|||'--> dejam counter (1)
||'---> ignition counter (2)
|'----> substate (0=no substate)
'-----> main state (2 = ignition)
```
When exposing the counters individually, same argumentation as above...

Regarding the counter-1-digit-limitation:
This, however, would mean that we limit the max. attempts of ignition and dejam to one digit (which should be fine).
Regarding this we should keep in mind the real-world cases:
* dejamming: The datasheet says attempts can be made after "approx. 60s". Obviously we would like it to be as short as possible. Since I do not know what affects the exact timing (may ambient temperature?), I added multiple dejamming attempts in the first place. Could be irrelevant once we tested and minimized the timers with hardware. --> unlikely we need more than 9 attempts.
* ignition: Most likely scenario is that after setting up the hardware, we forgot to open the manual gas valve. Hence an additional attempt would be nice. The "approx. 60 s" should give us enough time to open the valve --> 2 attempts are likely to be enough. Other case: Its very windy and therefore ignition does not work reliably --> if more than 9 attempts are required, it is bedder to use a less windy spot or windshield rather than increasing the ignition attempts.

I will push a draft shortly.

Reply:
I've pushed a draft: 92f99ea.